## Table of Contents

1) Import Libraries and Data

2) Combine orders/products data with products data

3) Check results of merge

4) Export dataframe

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#define path
path = r'C:\Users\Owner\Documents\Career Foundry\Tasks\Data Immersion Tasks\Instacart Project\2 Data'

In [3]:
#import orders_products_combined pkl file from 4.6 exercises as df_ords_prods
df_ords_prods = pd.read_pickle(os.path.join(path, 'prepared data', 'orders_products_combined.pkl'))

In [6]:
#check shape of df_ords_prods to make sure it matches (32434489, 10)
df_ords_prods.shape

(32434489, 10)

## 2) 
Determine a suitable way to combine the orders_products_combined dataframe with your products data set. Make sure you’re using your wrangled, cleaned, and deduped products data set stored in your “Prepared Data” folder from the previous Exercise’s task.

In [13]:
#import cleaned and deduped product data from prepared folder as df_prods using unamed column as index
df_prods = pd.read_csv(os.path.join(path, 'prepared data', 'products_cleaned.csv'), index_col = 'Unnamed: 0')

In [14]:
df_prods.shape

(49672, 5)

In [ ]:
#compare the head results of df_ords and df_ords_prods to look for a column match

In [15]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [16]:
df_ords_prods.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [19]:
#remove the merge column from df_ords_prods
df_ords_prods = df_ords_prods.drop(columns =['_merge'])

In [22]:
#df_prods and df_ords_prods both have a product_id column. will merge on that with outer join and merge flag count 
df_ords_prods_total = df_ords_prods.merge(df_prods, on ='product_id', how = 'outer', indicator = True)

## 3)
Confirm the results of the merge using the merge flag.

In [23]:
#check shape of df_ords_prods_total
df_ords_prods_total.shape

(32435070, 14)

In [24]:
df_ords_prods_total.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                1.0        0.0   
1                    15.0         196                1.0        1.0   
2                    21.0         196                1.0        1.0   
3                    29.0         196                1.0        1.0   
4                    28.0         196                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  
0         Soda      77.0            7.0     9.0   both  
1         Soda      77.0            7.0     9.0   both  
2         Soda      77.0            7.0     9.0   both  
3         Soda      77.0            7.0     9.0   both  
4         Soda      77.0            7.0     9.0   both

In [25]:
#count the values of df_ords_prods_total merge column to check results of merge
df_ords_prods_total['_merge'].value_counts()

both          32404859
left_only        30200
right_only          11
Name: _merge, dtype: int64

In [30]:
#create subset of the 11 values in df_prods that were not in df_ords_prods
df_prods_11 = df_ords_prods_total[df_ords_prods_total['_merge']=='right_only']

In [31]:
df_prods_11

order_id  user_id  order_number  orders_day_of_week  \
32435059       NaN      NaN           NaN                 NaN   
32435060       NaN      NaN           NaN                 NaN   
32435061       NaN      NaN           NaN                 NaN   
32435062       NaN      NaN           NaN                 NaN   
32435063       NaN      NaN           NaN                 NaN   
32435064       NaN      NaN           NaN                 NaN   
32435065       NaN      NaN           NaN                 NaN   
32435066       NaN      NaN           NaN                 NaN   
32435067       NaN      NaN           NaN                 NaN   
32435068       NaN      NaN           NaN                 NaN   
32435069       NaN      NaN           NaN                 NaN   

          order_hour_of_day  days_since_prior_order  product_id  \
32435059                NaN                     NaN        3630   
32435060                NaN                     NaN        3718   
32435061                NaN                     NaN        7045   
32435062                NaN                     NaN       25383   
32435063                NaN                     NaN       27499   
32435064                NaN                     NaN       36233   
32435065                NaN                     NaN       37703   
32435066                NaN                     NaN       43725   
32435067                NaN                     NaN       45971   
32435068                NaN                     NaN       46625   
32435069                NaN                     NaN       49540   

          add_to_cart_order  reordered  \
32435059                NaN        NaN   
32435060                NaN        NaN   
32435061                NaN        NaN   
32435062                NaN        NaN   
32435063                NaN        NaN   
32435064                NaN        NaN   
32435065                NaN        NaN   
32435066                NaN        NaN   
32435067                NaN        NaN   
32435068                NaN        NaN   
32435069                NaN        NaN   

                                               product_name  aisle_id  \
32435059                        Protein Granola Apple Crisp      57.0   
32435060                   Wasabi Cheddar Spreadable Cheese      21.0   
32435061             Unpeeled Apricot Halves in Heavy Syrup      88.0   
32435062                                 Chocolate Go Bites      61.0   
32435063                  Non-Dairy Coconut Seven Layer Bar     100.0   
32435064                            Water With Electrolytes     100.0   
32435065  Ultra Sun Blossom Liquid 90 loads Fabric Enhan...      75.0   
32435066                               Sweetart Jelly Beans     100.0   
32435067                         12 Inch Taper Candle White     101.0   
32435068    Single Barrel Kentucky Straight Bourbon Whiskey      31.0   
32435069                            Pure Squeezed  Lemonade      31.0   

          department_id  prices      _merge  
32435059           14.0     2.6  right_only  
32435060           16.0    12.0  right_only  
32435061           13.0     6.8  right_only  
32435062           19.0     1.8  right_only  
32435063           21.0    15.0  right_only  
32435064           21.0    10.4  right_only  
32435065           17.0    14.3  right_only  
32435066           21.0     8.1  right_only  
32435067           17.0     9.8  right_only  
32435068            7.0     1.7  right_only  
32435069            7.0     9.1  right_only

In [34]:
# testing to make sure product a id (3630) really doesnt exist in df_ords_prods
df_ords_prods[df_ords_prods['product_id']== 3630]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered]
Index: []

In [35]:
#create subset of the values in df_ords_prods that were not in df_prods
df_ords_prods_30200 = df_ords_prods_total[df_ords_prods_total['_merge']=='left_only']

In [38]:
df_ords_prods_30200.head()

order_id  user_id  order_number  orders_day_of_week  \
12039117     7099.0     27.0          63.0                 3.0   
12039118  1837192.0     27.0          80.0                 2.0   
12039119  2915432.0    298.0           1.0                 4.0   
12039120   613874.0    298.0           3.0                 1.0   
12039121   690386.0    479.0           1.0                 3.0   

          order_hour_of_day  days_since_prior_order  product_id  \
12039117               10.0                     1.0        6799   
12039118                8.0                     6.0        6799   
12039119               14.0                     NaN        6799   
12039120               12.0                    14.0        6799   
12039121               17.0                     NaN        6799   

          add_to_cart_order  reordered product_name  aisle_id  department_id  \
12039117                1.0        0.0          NaN       NaN            NaN   
12039118                9.0        1.0          NaN       NaN            NaN   
12039119                2.0        0.0          NaN       NaN            NaN   
12039120                2.0        1.0          NaN       NaN            NaN   
12039121                7.0        0.0          NaN       NaN            NaN   

          prices     _merge  
12039117     NaN  left_only  
12039118     NaN  left_only  
12039119     NaN  left_only  
12039120     NaN  left_only  
12039121     NaN  left_only

In [41]:
# testing to make sure product a id (6799) really doesnt exist in df_prods
df_prods[df_prods['product_id']== 6799]

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

### Merge Results
32404859 rows with product id shared among both sets

30200 rows from combined product and order data df_ords_prods that have product ids not present in the products data

11 rows of product data df_prods that have product ids not in df_ords_prods

I am not sure what analysis will be done with this so I will keep the outer join data and exclude nulls later if necessary.

## 4
Export the newly created dataframe as orders_products_merged in a suitable format (taking into consideration the size).

In [42]:
#export df_ords_prods_total as orders_products_merged using pickle format
df_ords_prods_total.to_pickle(os.path.join(path, 'prepared data', 'orders_products_merged.pkl'))